gitless/core.py

```diff
@@ -502,20 +502,31 @@ def listall_branches(self):
       yield regex.match(head).group(1)
 
   def lookup_branch(self, branch_name):
-    if not stdout(git('ls-remote', '--heads', self.name, branch_name)):
+    branches = self.lookup_branches([branch_name])
+    if not branches:
       return None
-    # The branch exists in the remote
-    git.fetch(self.git_remote.name, branch_name)
-    git_branch = self.gl_repo.git_repo.lookup_branch(
-        self.git_remote.name + '/' + branch_name, pygit2.GIT_BRANCH_REMOTE)
-    # Make another check for the branch being None
-    # As observed in issue : https://github.com/sdg-mit/gitless/issues/211
-    if git_branch is None:
-        git.fetch(self.git_remote.name)
-        git_branch = self.gl_repo.git_repo.lookup_branch(
-            self.git_remote.name + '/' + branch_name, pygit2.GIT_BRANCH_REMOTE)
-    return RemoteBranch(git_branch, self.gl_repo)
-
+    else:
+      return branches[0]
+
+  def lookup_branches(self, branch_names):
+    all_branches = self.listall_branches()
+    branch_names = [branch_name for branch_name in all_branches 
```

Comment:
This will return them unsorted when it's probably better to respect the sort of the argument `branch_names` (if otherwise the `sorted(..)` in gl.remote is a no-op)

Perhaps you should have a `lookupall_branches` that will internally call
`lookup_branches(sorted(self.listall_branches())`. That way in gl.remote we don't end up calling `listall_branches` twice. And you can change the behavior of `lookup_branches` to not do the check: just let it throw an error on `git.fetch` or if you want you can catch the error and return `None`.

I am looking at the uses of `lookup_branch` and I think that here:

https://github.com/sdg-mit/gitless/blob/d2620f92f04229f4d29e5d7399d34fc59dbfe349/gitless/core.py#L146

we need to also catch `AttributeError` in case `lookup_branch` returns `None`